File: rxbinding-recyclerview-v7/src/main/java/com/jakewharton/rxbinding/support/v7/widget/RxRecyclerViewAdapter.java
package com.jakewharton.rxbinding.support.v7.widget;

import android.support.annotation.CheckResult;
import android.support.v7.widget.RecyclerView;

import rx.Observable;

/**
 * Static factory methods for creating {@linkplain Observable observables} for {@link RecyclerView.Adapter}.
 */
public class RxRecyclerViewAdapter {

    /**
     * Create an observable of data change events for {@code RecyclerView.adapter}.
     * <p>
     * <em>Note:</em> A value will be emitted immediately on subscribe.
     */
    @CheckResult
    public static <T extends RecyclerView.Adapter> Observable<T> dataChanges(T adapter) {
        return Observable.create(new RecyclerAdapterDataChangeOnSubscribe<>(adapter));
    }

    private RxRecyclerViewAdapter() {
        throw new AssertionError("No instances.");
    }
}

File: rxbinding-recyclerview-v7-kotlin/src/main/kotlin/com/jakewharton/rxbinding/support/v7/widget/RxRecyclerViewAdapter.kt
package com.jakewharton.rxbinding.support.v7.widget

import android.support.v7.widget.RecyclerView
import com.jakewharton.rxbinding.support.v7.widget.RxRecyclerViewAdapter
import android.support.v7.widget.RecyclerView.Adapter
import rx.Observable

/**
 * Create an observable of data change events for recyclerview `adapter`.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
public inline fun <T : Adapter<out RecyclerView.ViewHolder>> T.dataChanges(): Observable<T> = RxRecyclerViewAdapter.dataChanges(this)
